python/hail/expr.py

```diff
@@ -26,6 +27,7 @@ class Type(object):
 
     def __init__(self, jtype):
         self._jtype = jtype
+        self._history = None
```

Comment:
I think it is unnecessary to track history on `Type`.  You should just define `__repr__` to return the python code necessary to recreate the type.  This may give you an opportunity to simplify the record methods.

The one case where I could imagine this being useful is something like:

```
vds = hc.read(...).filter_samples_expr(...)
hc.import_table(..., types={'Col': vds.variant_schema}).write(...)
```

Do you want to include the pipeline that produces `vds` in the history for the table write for the `types` argument?